### PR TITLE
#923 Make MasterSlaveReplication iterable by extending AbstractConnection

### LINKED ISSUE #923: Error while iterating on `Predis\Client`: `Object of class Predis\Connection\Aggregate\MasterSlaveReplication could not be converted to string`
**Describe the bug**
I get a fatal error when iterating on a `Predis\Client` instance which connection is a `Predis\Connection\Aggregate\MasterSlaveReplication` instance.

**To Reproduce**
With `snc/redis-bundle 4.4.2` and `predis/predis 1.1.10`, `Symfony 5.4.16`, and `php 8.1.11` and the following config:

```yaml
snc_redis:
    clients:
        default:
            type: predis
            alias: default
            dsn:
                - "%env(REDIS_URL)%?alias=master"
                - "%env(REDIS_URL_READ_ONLY)%"
            logging: '%kernel.debug%'
            options:
                replication: true
```

Then run:

```
$ bin/console redis:query flushall -vvv

In Client.php line 540:

  [Error]
  Object of class Predis\Connection\Aggregate\MasterSlaveReplication could not be converted to string


Exception trace:
  at /srv/app/vendor/predis/predis/src/Client.php:540
 Predis\Client->getIterator() at /srv/app/vendor/snc/redis-bundle/src/Command/RedisQueryCommand.php:86
 Snc\RedisBundle\Command\RedisQueryCommand->execute() at /srv/app/vendor/symfony/console/Command/Command.php:312
 Symfony\Component\Console\Command\Command->run() at /srv/app/vendor/symfony/console/Application.php:1038
 Symfony\Component\Console\Application->doRunCommand() at /srv/app/vendor/symfony/framework-bundle/Console/Application.php:96
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /srv/app/vendor/symfony/console/Application.php:312
 Symfony\Component\Console\Application->doRun() at /srv/app/vendor/symfony/framework-bundle/Console/Application.php:82
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /srv/app/vendor/symfony/console/Application.php:168
 Symfony\Component\Console\Application->run() at /srv/app/bin/console:43

redis:query [--client CLIENT] [--] <query>...
```

**Expected behavior**
It should output:

```
DEBUG     [snc_redis] Executing command "FLUSHALL"
^ Predis\Response\Status^ {#389
  -payload: "OK"
}
```

**Versions (please complete the following information):**
 - Predis: 1.1.10
 - snc/redis-bundle 4.4.2
 - Symfony 5.4.16
 - PHP 8.1.11
 - Redis Server: bitnami/redis:6.2.6

**Code sample**
The code that tries to iterate on the client is [here](https://github.com/snc/SncRedisBundle/blob/master/src/Command/RedisQueryCommand.php#L86).

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Gregoire-M**
Fixes #923

### CODE REVIEW

File: src/Connection/Replication/MasterSlaveReplication.php [hunk 1]
```diff
@@ -77,6 +79,34 @@ public function __construct(ReplicationStrategy $strategy = null)
...
+    {
+        throw new \RuntimeException('Not implemented.');
+    }
+
+    public function read()
+    {
+        throw new \RuntimeException('Not implemented.');
+    }
+
+    protected function getIdentifier()
```
Comment (tillkruss) [inline src/Connection/Replication/MasterSlaveReplication.php:97]: We need a test for this on the `MasterSlaveReplication`.

File: tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php [hunk 2]
```diff
@@ -268,12 +268,12 @@ public function testDisconnectForcesCurrentConnectionToDisconnect(): void
...
         $master = $this->getMockConnection('tcp://127.0.0.1:6379?role=master');
         $master
-            ->expects($this->once())
+            ->expects($this->exactly(2))
             ->method('disconnect');
 
         $slave = $this->getMockConnection('tcp://127.0.0.1:6380?role=slave');
         $slave
-            ->expects($this->once())
+            ->expects($this->exactly(2))
```
Comment (tillkruss) [inline tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php:276]: Why is this changing?

**Review by tillkruss — CHANGES_REQUESTED** [review]

File: src/Connection/Replication/MasterSlaveReplication.php [hunk 3]
```diff
@@ -77,6 +79,34 @@ public function __construct(ReplicationStrategy $strategy = null)
...
     }
 
+    protected function assertParameters(ParametersInterface $parameters)
+    {
+        return $parameters;
+    }
+
+    protected function createResource()
+    {
+        throw new \RuntimeException('Not implemented.');
```
Comment (vladvildanov) [inline src/Connection/Replication/MasterSlaveReplication.php:89]: MInor: Use FQDN

File: tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php (hunk 2 above)
Comment (vladvildanov) [reply tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php:276]: Same question, why `disconnect` should be called twice per connection?

**Review by vladvildanov — CHANGES_REQUESTED** [review]

File: tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php (hunk 2 above)
Comment (Gregoire-M) [reply tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php:276]: Now it's also called in `AbstractConnection` destructor

Comment (tillkruss) [reply tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php:276]: We can't call disconnect twice.

File: tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php [hunk 4]
```diff
@@ -268,12 +268,12 @@ public function testDisconnectForcesCurrentConnectionToDisconnect(): void
     {
         $master = $this->getMockConnection('tcp://127.0.0.1:6379?role=master');
         $master
-            ->expects($this->once())
+            ->expects($this->exactly(2))
```
Comment (tillkruss) [inline tests/Predis/Connection/Replication/MasterSlaveReplicationTest.php:271]: We can't call disconnect twice.

**Review by tillkruss — CHANGES_REQUESTED** [review]

**Pull request closed without merging**